# 02 — Preprocessing: EyePACS and OLIVES

## Purpose

Preprocess both source datasets into normalised 224×224 PyTorch tensors saved to Google Drive, ready for downstream feature extraction and model training. This notebook captures the **complete preprocessing pipeline** including the iterative debugging process that surfaced three substantive bugs.

## Two distinct phases of this notebook

**Cells 1-12 — Production preprocessing flow.** Clean, repeatable preprocessing of both datasets. Run these cells in order in a fresh Colab session to reproduce the final preprocessed datasets.

**Cells 13-51 — Debugging investigation.** Diagnostic and remediation cells from the iterative process of fixing three preprocessing bugs:
1. Clinical xlsx column name mismatch (`File_Path` vs the biomarker CSV's path column)
2. Biomarker deduplication losing positives (`drop_duplicates(keep='first')` → `groupby().max()`)
3. Prime_FULL `.tif`/`.png` duplication (same fundus image stored in both formats)

These are retained as historical evidence of methodological rigour — see `docs/PHASE1_SUMMARY.md` Section 6 for the full narrative.

## Inputs

- EyePACS HuggingFace dataset at `MyDrive/` (arrow files)
- `olive.zip` (33.92 GB) on `MyDrive/` containing the nested OLIVES archives
- `OLIVES_Dataset_Labels.zip` on `MyDrive/` (already extracted by `01c_data_inspection.ipynb`)
- GitHub token stored in Colab Secrets as `GITHUB_TOKEN`

## Outputs (saved to Drive)

- `dissertation/preprocessed/eyepacs/eyepacs_shard_*.pt` (8 shards, ~22 GB total) — 35,108 fundus images
- `dissertation/preprocessed/olives/olives_fundus.pt` (~1.8 GB) — 3,142 unique fundus images with three-tier labels and masks
- Two `metadata.json` files describing each dataset's preprocessing parameters

## Runtime requirements

- T4 GPU minimum (Runtime → Change runtime type → T4 GPU)
- Full preprocessing run: ~30-60 minutes for EyePACS, ~1-2 hours for OLIVES (one-time cost)
- Resumable: existing shards/outputs are detected and skipped on re-run


In [ ]:
# ── PHASE 2 | CELL 1 ── Session Setup
# PURPOSE: Mounts Google Drive, clones or pulls the latest
# repo from GitHub, and sets the Python path.
# RUN: Every session — nothing works without this.

import os
import sys
from google.colab import drive

# Setup: mount Drive, clone the public repo
drive.mount('/content/drive')

REPO_DIR = "/content/dr-dissertation"
if not os.path.exists(REPO_DIR):
    !git clone https://github.com/savita10/dr-dissertation.git {REPO_DIR}
else:
    !git -C {REPO_DIR} fetch origin && git -C {REPO_DIR} reset --hard origin/main

import torch
print(f"CUDA available: {torch.cuda.is_available()}")
if torch.cuda.is_available():
    print(f"GPU: {torch.cuda.get_device_name(0)}")
else:
    print("WARNING: no GPU — set Runtime → Change runtime type → GPU")

# Move into repo and add to Python path
os.chdir('/content/dr-dissertation')
sys.path.insert(0, '/content/dr-dissertation')
print(f"Working directory: {os.getcwd()}")

Mounted at /content/drive
Drive mounted
Cloning into 'dr-dissertation'...
remote: Enumerating objects: 93, done.
remote: Counting objects: 100% (93/93), done.
remote: Compressing objects: 100% (59/59), done.
remote: Total 93 (delta 42), reused 78 (delta 27), pack-reused 0 (from 0)
Receiving objects: 100% (93/93), 27.81 KiB | 1.07 MiB/s, done.
Resolving deltas: 100% (42/42), done.
Repo cloned
Working directory: /content/dr-dissertation


In [ ]:
# ── PHASE 2 | CELL 2 ── Install PyTorch for Blackwell GPU (cu128)
# PURPOSE: Installs PyTorch 2.7.0 with CUDA 12.8 support,
# required for the Blackwell GPU (sm_120 architecture).
# RUN: Every session — MUST run before any 'import torch'.
# NOTE: Dependency conflict warnings at the end are harmless.
!pip install torch==2.7.0 torchvision==0.22.0 \
    --index-url https://download.pytorch.org/whl/cu128 \
    --force-reinstall -q
print("PyTorch installed")

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 88.0/88.0 MB 101.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 954.8/954.8 kB 128.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.2/10.2 MB 161.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 726.9/726.9 MB 62.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 609.6/609.6 MB 65.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 193.1/193.1 MB 121.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 63.6/63.6 MB 144.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 260.4/260.4 MB 75.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 292.1/292.1 MB 108.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 156.8/156.8 MB 134.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 201.3/201.3 MB 4.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 89.9/89.9 kB 84.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━

In [ ]:
# ── PHASE 2 | CELL 3 ── Install Project Requirements
# PURPOSE: Installs timm, omegaconf, datasets and other
# packages needed by src/data/preprocess.py.
# RUN: Every session.
# NOTE: Harmless warnings expected — check last line only.
!pip install -r requirements.txt -q
!pip install timm omegaconf datasets -q
print("Requirements installed")

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.9/60.9 kB 5.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 193.6/193.6 kB 20.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 19.2/19.2 MB 61.8 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
torchaudio 2.10.0+cu128 requires torch==2.10.0, but you have torch 2.7.0+cu128 which is incompatible.
gradio 5.50.0 requires pillow<12.0,>=8.0, but you have pillow 12.1.1 which is incompatible.
Requirements installed


In [ ]:
# ── PHASE 2 | CELL 4 ── Verify GPU
# PURPOSE: Confirms GPU is available and PyTorch is using
# the correct CUDA version.
# RUN: Every session — if this fails, stop and restart runtime.
# Do NOT skip — broken GPU means 50x slower preprocessing.
import torch
print(f"PyTorch version: {torch.__version__}")
print(f"CUDA available: {torch.cuda.is_available()}")
if torch.cuda.is_available():
    print(f"GPU: {torch.cuda.get_device_name(0)}")
    print(f"CUDA version: {torch.version.cuda}")
    x = torch.ones(3, 3).cuda()
    print(f"GPU test: {(x*2)[0,0].item()} (expected 2.0)")
    print("GPU working correctly")
else:
    print("WARNING: GPU not available")

PyTorch version: 2.7.0+cu128
CUDA available: True
GPU: Tesla T4
CUDA version: 12.8
GPU test: 2.0 (expected 2.0)
GPU working correctly


In [ ]:
# ── PHASE 2 | CELL 5 ── Project Setup and Drive Paths
# PURPOSE: Runs setup_colab() which creates output directories
# on Drive and returns the paths dict used by all later cells.
# RUN: Every session — 'paths' variable is required by Cells 9-11.
from src.utils.colab_setup import setup_colab
paths = setup_colab()
print("Setup complete")
print("Paths:", paths)

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
GPU: Tesla T4 (14.6 GB)
RAM: 12.7 GB total
Disk (/content): 185.0 GB free / 235.7 GB total
Setup complete
Paths: {'drive_root': '/content/drive/MyDrive/dissertation', 'raw_data_dir': '/content/drive/MyDrive/dissertation/raw', 'preprocessed_dir': '/content/drive/MyDrive/dissertation/preprocessed', 'features_dir': '/content/drive/MyDrive/dissertation/features', 'results_dir': '/content/drive/MyDrive/dissertation/results', 'local_cache': '/content/cache', 'local_scratch': '/content/scratch', 'eyepacs_hf_dir': '/content/drive/MyDrive', 'olives_labels_dir': '/content/drive/MyDrive/olives_labels/', 'olives_extracted_dir': '/content/drive/MyDrive/dissertation/raw/olives_extracted/', 'eyepacs_zip': '/content/drive/MyDrive/EYEPACS.zip', 'olive_zip': '/content/drive/MyDrive/olive.zip', 'olives_labels_csv': '/content/drive/MyDrive/olives_labels/OLIVES_Dataset_Labels/ml_

In [ ]:
# ── PHASE 2 | CELL 6 ── Create Local Scratch Directories
# PURPOSE: Creates /content/scratch on Colab's local SSD.
# Used by OLIVES preprocessing for fast temporary extraction.
# RUN: Every session — wiped when session ends, must be recreated.
import os
os.makedirs('/content/scratch', exist_ok=True)
os.makedirs('/content/scratch/olives_fundus', exist_ok=True)
print("Scratch directories created:")
print("  /content/scratch")
print("  /content/scratch/olives_fundus")

Scratch directories created:
  /content/scratch
  /content/scratch/olives_fundus


In [ ]:
import os
os.makedirs('/content/eyepacs_hf', exist_ok=True)
print("Created /content/eyepacs_hf")

Created /content/eyepacs_hf


In [ ]:
# ── PHASE 2 | CELL 6b ── Copy EyePACS to Local SSD
# PURPOSE: Copies the 14 HuggingFace arrow files + JSON files
# from MyDrive root to /content/eyepacs_hf on local SSD.
# RUN: Every session before running Cell 7.
# WHY: Reading directly from Drive causes OOM kills.
#      Local SSD is 10x faster and avoids throttling.
# TIME: 3-5 minutes per session.
import os
import shutil

# Create destination directory
os.makedirs('/content/eyepacs_hf', exist_ok=True)  # ← add this line

drive_root = '/content/drive/MyDrive'

# Scan for HuggingFace files
print("Scanning MyDrive root for HuggingFace files...")
src = '/content/drive/MyDrive'
dst = '/content/eyepacs_hf'

# Check what HuggingFace files exist in MyDrive root
print("Scanning MyDrive root for HuggingFace files...")
all_files = os.listdir(src)
arrow_files = [f for f in all_files if f.endswith('.arrow')]
has_info = 'dataset_info.json' in all_files
has_state = 'state.json' in all_files

print(f"  Arrow files found: {len(arrow_files)}")
print(f"  dataset_info.json: {has_info}")
print(f"  state.json: {has_state}")

if len(arrow_files) == 0:
    print("\nERROR: No arrow files found in MyDrive root")
    print("Check your Drive structure — files may be in a subfolder")
else:
    # Copy files one by one with progress
    files_to_copy = arrow_files.copy()
    if has_info:
        files_to_copy.append('dataset_info.json')
    if has_state:
        files_to_copy.append('state.json')

    print(f"\nCopying {len(files_to_copy)} files to {dst}...")
    for f in files_to_copy:
        src_path = os.path.join(src, f)
        dst_path = os.path.join(dst, f)
        size_mb = os.path.getsize(src_path) / 1e6
        print(f"  Copying {f} ({size_mb:.0f} MB)...", end=' ')
        shutil.copy2(src_path, dst_path)
        print("done")

    # Verify
    copied = os.listdir(dst)
    print(f"\nVerification — files in {dst}:")
    for f in sorted(copied):
        size = os.path.getsize(os.path.join(dst, f)) / 1e6
        print(f"  {f} ({size:.1f} MB)")

Scanning MyDrive root for HuggingFace files...
Scanning MyDrive root for HuggingFace files...
  Arrow files found: 14
  dataset_info.json: True
  state.json: True

Copying 16 files to /content/eyepacs_hf...
  Copying data-00000-of-00014.arrow (466 MB)... done
  Copying data-00001-of-00014.arrow (467 MB)... done
  Copying data-00002-of-00014.arrow (466 MB)... done
  Copying data-00003-of-00014.arrow (465 MB)... done
  Copying data-00004-of-00014.arrow (471 MB)... done
  Copying data-00005-of-00014.arrow (468 MB)... done
  Copying data-00006-of-00014.arrow (467 MB)... done
  Copying data-00007-of-00014.arrow (467 MB)... done
  Copying data-00008-of-00014.arrow (467 MB)... done
  Copying data-00009-of-00014.arrow (466 MB)... done
  Copying data-00010-of-00014.arrow (469 MB)... done
  Copying data-00011-of-00014.arrow (465 MB)... done
  Copying data-00012-of-00014.arrow (466 MB)... done
  Copying data-00013-of-00014.arrow (469 MB)... done
  Copying dataset_info.json (0 MB)... done
  Copyin

In [ ]:

# ── PHASE 2 | CELL 7 ── Preprocess EyePACS Fundus Images
# PURPOSE: Loads EyePACS from local SSD, resizes to 224x224,
# normalises with ImageNet stats, saves as 8 sharded .pt files.
# STATUS: ✅ COMPLETE — 35,108 images saved to Drive.
# RUN: Skip — output already on Drive.
#      Only rerun if shards are deleted or corrupted.
# OUTPUT: dissertation/preprocessed/eyepacs/ (8 shards + metadata.json)

import subprocess
import sys
import os

# First check if preprocessing already completed
eyepacs_out = '/content/drive/MyDrive/dissertation/preprocessed/eyepacs'
meta_path = f'{eyepacs_out}/metadata.json'

if os.path.exists(meta_path):
    import json
    with open(meta_path) as f:
        meta = json.load(f)
    print("EyePACS preprocessing already complete!")
    print(f"Total images: {meta.get('total_images', 'unknown')}")
    print(f"Shards: {meta.get('num_shards', 'unknown')}")
    print("Skipping — delete metadata.json to force rerun")
else:
    print("Starting EyePACS preprocessing...")
    print("Expected time: 30-60 minutes")
    print("Keep this browser tab active to prevent disconnection")
    print("="*50)

    env = os.environ.copy()
    env['PYTHONPATH'] = '/content/dr-dissertation'

    result = subprocess.run(
        [sys.executable,
         '/content/dr-dissertation/src/data/preprocess.py',
         '--dataset', 'eyepacs',
         '--config',
         '/content/dr-dissertation/configs/preprocess.yaml'],
        capture_output=True,
        text=True,
        env=env
    )

    print("=== STDOUT ===")
    print(result.stdout if result.stdout else "(no stdout)")
    print("\n=== STDERR ===")
    print(result.stderr if result.stderr else "(no stderr)")
    print(f"\nExit code: {result.returncode}")
    if result.returncode == 0:
        print("EyePACS preprocessing complete")
    else:
        print("ERROR — check output above")

EyePACS preprocessing already complete!
Total images: 35108
Shards: 8
Skipping — delete metadata.json to force rerun


In [ ]:
# ── PHASE 2 | CELL 8 ── Preprocess OLIVES Fundus Images
# PURPOSE: Extracts fundus images from triple-nested zip
# (olive.zip → OLIVES.zip → TREX_DME.zip / Prime_FULL.zip),
# joins to biomarker CSV, preprocesses and saves to Drive.
#
# RUNTIME: 1-2 hours (nested zip extraction is slow)
# RESUMABLE: Yes — skips already processed files
# RUN ONCE: After completion skip this cell in future sessions
# WARNING: Run Cell 7 first and confirm it completed
#
# OUTPUT: dissertation/preprocessed/olives/
#   olives_fundus.pt
#   metadata.json

import subprocess
import sys
import os

olives_out = '/content/drive/MyDrive/dissertation/preprocessed/olives'
meta_path = f'{olives_out}/metadata.json'

if os.path.exists(meta_path):
    import json
    with open(meta_path) as f:
        meta = json.load(f)
    print("OLIVES preprocessing already complete!")
    print(f"Total images: {meta.get('total_images', 'unknown')}")
    print("Skipping — delete metadata.json to force rerun")
else:
    print("Starting OLIVES preprocessing...")
    print("Expected time: 1-2 hours")
    print("Keep this browser tab active")
    print("="*50)

    env = os.environ.copy()
    env['PYTHONPATH'] = '/content/dr-dissertation'

    result = subprocess.run(
        [sys.executable,
         '/content/dr-dissertation/src/data/preprocess.py',
         '--dataset', 'olives',
         '--config',
         '/content/dr-dissertation/configs/preprocess.yaml'],
        capture_output=True,
        text=True,
        env=env
    )

    print("=== STDOUT ===")
    print(result.stdout if result.stdout else "(no stdout)")
    print("\n=== STDERR ===")
    print(result.stderr if result.stderr else "(no stderr)")
    print(f"\nExit code: {result.returncode}")
    if result.returncode == 0:
        print("OLIVES preprocessing complete")
    else:
        print("ERROR — check output above")

Starting OLIVES preprocessing...
Expected time: 1-2 hours
Keep this browser tab active
=== STDOUT ===

OLIVES preprocessing

[1/5 Extracting fundus images to /content/scratch/olives_fundus]
  Opening outer zip: /content/drive/MyDrive/olive.zip
  Opening OLIVES.zip as stream...
  Processing OLIVES/TREX_DME.zip...
    Found 1849 fundus files in TREX_DME
  Processing OLIVES/Prime_FULL.zip...
    Found 2586 fundus files in Prime_FULL
extraction complete in 1805.2s

[2/5 Loading labels from /content/drive/MyDrive/olives_labels/OLIVES_Dataset_Labels/ml_centric_labels/Biomarker_Clinical_Data_Images.csv]
loaded 9408 rows, 192 unique (trial, patient, visit, eye) keys in 0.1s

[3/5 Walking scratch dir for fundus files]
found 4435 fundus files on disk

[4/5 Preprocessing matched images]
processed=272, unmatched=4163, failed=0 in 5.9s

[5/5 Saving tensor file and metadata.json]
saved in 1.0s — output: /content/drive/MyDrive/dissertation/preprocessed/olives/olives_fundus.pt

[DONE] olives preproces

In [ ]:
# ── PHASE 2 | CELL 9 ── Verify EyePACS Preprocessing Output
# PURPOSE: Loads shard 0 and confirms shape, dtype, normalisation.
# STATUS: ✅ Already verified — 8 shards, normalisation PASSED.
# RUN: Skip — already confirmed correct.
#      Only rerun if you suspect shard corruption.
import torch, json, os

eyepacs_dir = '/content/drive/MyDrive/dissertation/preprocessed/eyepacs/'

# Load and print metadata
with open(eyepacs_dir + 'metadata.json') as f:
    meta = json.load(f)
print("EyePACS metadata:")
for k, v in meta.items():
    print(f"  {k}: {v}")

# Verify shards
shards = sorted([f for f in os.listdir(eyepacs_dir) if f.endswith('.pt')])
print(f"\nShards found: {len(shards)}")

# Check first shard
shard = torch.load(eyepacs_dir + shards[0], map_location='cpu')
print(f"\nFirst shard:")
print(f"  images shape: {shard['images'].shape}")
print(f"  labels shape: {shard['labels'].shape}")
print(f"  normalisation check: "
      f"{'PASSED' if shard['images'].min() > -3 and shard['images'].max() < 3 else 'FAILED'}")

EyePACS metadata:
  total_images: 35108
  num_shards: 8
  image_size: [224, 224]
  normalisation: {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}
  class_distribution: {'0': 25802, '1': 2438, '2': 5288, '3': 872, '4': 708}
  preprocessing_date: 2026-04-26T11:28:38.126998+00:00

Shards found: 8

First shard:
  images shape: torch.Size([5000, 3, 224, 224])
  labels shape: torch.Size([5000])
  normalisation check: PASSED


In [ ]:
# ── PHASE 2 | CELL 10 ── Verify OLIVES Preprocessing Output
# PURPOSE: Loads olives_fundus.pt and confirms all tensors:
# images, labels, biomarkers, bcva, cst shapes and values.
# RUN: Once after Cell 8 completes successfully.
# SKIP: In future sessions after verification passes.
import torch
import json
import os

olives_dir = paths['preprocessed_dir'] + '/olives/'

# Load metadata
meta_path = olives_dir + 'metadata.json'
if os.path.exists(meta_path):
    with open(meta_path) as f:
        meta = json.load(f)
    print("OLIVES metadata:")
    for k, v in meta.items():
        print(f"  {k}: {v}")
else:
    print("metadata.json not found — check preprocessing ran")

# Load OLIVES .pt file
pt_files = [
    f for f in os.listdir(olives_dir)
    if f.endswith('.pt')
]
print(f"\nOLIVES .pt files found: {pt_files}")

if pt_files:
    data = torch.load(
        olives_dir + pt_files[0],
        map_location='cpu'
    )
    print(f"\nOLIVES data contents:")
    for key, val in data.items():
        if isinstance(val, torch.Tensor):
            print(f"  {key}: shape={val.shape}, "
                  f"dtype={val.dtype}")
        elif isinstance(val, list):
            print(f"  {key}: list of {len(val)} items")
            if val:
                print(f"    sample: {val[0]}")

    # Check biomarker tensor
    if 'biomarkers' in data:
        bio = data['biomarkers']
        print(f"\nBiomarker check:")
        print(f"  Shape: {bio.shape} "
              f"(expected ~4435 x 16)")
        print(f"  Unique values: {bio.unique()}"
              f" (expected 0 and 1 only)")
        print(f"  Missing (NaN): "
              f"{torch.isnan(bio).sum().item()}")

OLIVES metadata:
  total_images: 272
  num_shards: 1
  image_size: [224, 224]
  normalisation: {'mean': [0.485, 0.456, 0.406], 'std': [0.229, 0.224, 0.225]}
  class_distribution: {'0': 272}
  preprocessing_date: 2026-04-26T20:34:40.590260+00:00

OLIVES .pt files found: ['olives_fundus.pt']

OLIVES data contents:
  images: shape=torch.Size([272, 3, 224, 224]), dtype=torch.float32
  labels: shape=torch.Size([272]), dtype=torch.int64
  metadata: list of 272 items
    sample: {'trial': 'TREX_DME', 'patient_id': '0213TOS', 'visit': 'V19', 'eye': 'OS'}
  biomarkers: shape=torch.Size([272, 16]), dtype=torch.float32
  bcva: shape=torch.Size([272]), dtype=torch.float32
  cst: shape=torch.Size([272]), dtype=torch.float32

Biomarker check:
  Shape: torch.Size([272, 16]) (expected ~4435 x 16)
  Unique values: tensor([0., 1., nan, nan, nan, nan, nan, nan, nan]) (expected 0 and 1 only)
  Missing (NaN): 7


In [ ]:
# ── PHASE 2 | CELL 11 ── Phase 2 Preprocessing Summary
# PURPOSE: Final confirmation both datasets are on Drive
# and ready for Phase 3 Distribution Validation.
# RUN: Once after both Cell 7 and Cell 8 complete.
# SKIP: In future sessions.

import os

eyepacs_dir = paths['preprocessed_dir'] + '/eyepacs/'
olives_dir = paths['preprocessed_dir'] + '/olives/'

eyepacs_shards = len([
    f for f in os.listdir(eyepacs_dir)
    if f.endswith('.pt')
]) if os.path.exists(eyepacs_dir) else 0

olives_files = len([
    f for f in os.listdir(olives_dir)
    if f.endswith('.pt')
]) if os.path.exists(olives_dir) else 0

print("=" * 50)
print("PHASE 2 PREPROCESSING SUMMARY")
print("=" * 50)
print(f"EyePACS shards saved:  {eyepacs_shards}")
print(f"OLIVES files saved:    {olives_files}")
print()
if eyepacs_shards > 0 and olives_files > 0:
    print("Both datasets preprocessed successfully")
    print("Ready to proceed to Phase 3:")
    print("Distribution Validation")
else:
    print("WARNING: One or both datasets missing")
    print("Re-run the failed preprocessing cell")
print("=" * 50)

PHASE 2 PREPROCESSING SUMMARY
EyePACS shards saved:  8
OLIVES files saved:    1

Both datasets preprocessed successfully
Ready to proceed to Phase 3:
Distribution Validation


---

# End of production preprocessing pipeline

The cells above (1-12) constitute the complete preprocessing flow. If both datasets are successfully produced, no further cells need to run.

The cells below are retained as historical evidence of the debugging process that found and fixed three preprocessing bugs. They are not part of the production flow — they were one-shot diagnostic and remediation cells executed during iterative development.


# Debugging session 1 — investigating low match count

**Symptom**: OLIVES preprocessing was matching only 272 out of 4,435 extracted fundus files to label rows.

**Hypothesis**: the key construction logic was generating keys that didn't align between the scratch fundus paths and the CSV path column.

Cells below progressively diagnose the mismatch, ultimately revealing two distinct bugs (CSV-vs-clinical-xlsx column name difference, and CSV vs scratch path tokenisation differences).


In [ ]:
import pandas as pd
from pathlib import Path

csv_path = '/content/drive/MyDrive/olives_labels/OLIVES_Dataset_Labels/ml_centric_labels/Biomarker_Clinical_Data_Images.csv'
df = pd.read_csv(csv_path)
PATH_COL = 'Path (Trial/Arm/Folder/Visit/Eye/Image Name)'

scratch_dir = Path('/content/scratch/olives_fundus')
fundus_files = [f for f in scratch_dir.rglob('*')
                if f.is_file() and 'fundus' in f.name.lower()]

# Build CSV keys both ways to see what works
def csv_key(path_str):
    parts = Path(path_str).parts
    trial = parts[1]
    if trial == 'Prime_FULL':
        return (parts[1], parts[2], parts[3], parts[4])
    else:  # TREX DME
        return (parts[1], parts[3], parts[4], parts[5])

csv_keys = set(df[PATH_COL].map(csv_key))

# Build scratch keys and check matches
matched, unmatched_samples = 0, []
for f in fundus_files:
    parts = f.relative_to(scratch_dir).parts
    trial = parts[0]
    if trial == 'Prime_FULL':
        key = (parts[0], parts[1], parts[2], parts[3])
    else:  # TREX DME
        key = (parts[0], parts[2], parts[3], parts[4])

    if key in csv_keys:
        matched += 1
    elif len(unmatched_samples) < 5:
        unmatched_samples.append((key, parts))

print(f"Total fundus files: {len(fundus_files)}")
print(f"Matched: {matched}")
print(f"Unmatched: {len(fundus_files) - matched}")
print(f"\nSample unmatched scratch keys:")
for key, parts in unmatched_samples:
    print(f"  key={key}")
    print(f"  parts={parts}")

print(f"\nSample CSV keys (Prime_FULL):")
prime_keys = [k for k in list(csv_keys)[:20] if k[0]=='Prime_FULL']
for k in prime_keys[:3]:
    print(f"  {k}")

print(f"\nSample CSV keys (TREX DME):")
trex_keys = [k for k in list(csv_keys)[:50] if k[0]=='TREX DME']
for k in trex_keys[:3]:
    print(f"  {k}")

Total fundus files: 0
Matched: 0
Unmatched: 0

Sample unmatched scratch keys:

Sample CSV keys (Prime_FULL):
  ('Prime_FULL', '01-027', 'W0', 'OS')
  ('Prime_FULL', '01-037', 'W12', 'OS')
  ('Prime_FULL', '02-046', 'W0', 'OD')

Sample CSV keys (TREX DME):
  ('TREX DME', '0248MOD', 'V26', 'OD')
  ('TREX DME', '0241GOD', 'V11', 'OD')
  ('TREX DME', '0215GOS', 'V12', 'OS')


In [ ]:
import os

meta_path = '/content/drive/MyDrive/dissertation/preprocessed/olives/metadata.json'
pt_path = '/content/drive/MyDrive/dissertation/preprocessed/olives/olives_fundus.pt'

if os.path.exists(meta_path):
    os.remove(meta_path)
    print("Deleted metadata.json")

if os.path.exists(pt_path):
    os.remove(pt_path)
    print("Deleted olives_fundus.pt")

Deleted metadata.json
Deleted olives_fundus.pt


In [ ]:
import os

meta = '/content/drive/MyDrive/dissertation/preprocessed/olives/metadata.json'
pt = '/content/drive/MyDrive/dissertation/preprocessed/olives/olives_fundus.pt'
scratch = '/content/scratch/olives_fundus'

print(f"metadata.json exists: {os.path.exists(meta)}")
print(f"olives_fundus.pt exists: {os.path.exists(pt)}")

# Check scratch
if os.path.exists(scratch):
    files = list(__import__('pathlib').Path(scratch).rglob('*'))
    fundus = [f for f in files if f.is_file() and 'fundus' in f.name.lower()]
    print(f"Scratch dir exists: True")
    print(f"Fundus files in scratch: {len(fundus)}")
else:
    print(f"Scratch dir exists: False")

metadata.json exists: True
olives_fundus.pt exists: True
Scratch dir exists: True
Fundus files in scratch: 4435


In [ ]:
import os

meta = '/content/drive/MyDrive/dissertation/preprocessed/olives/metadata.json'
pt = '/content/drive/MyDrive/dissertation/preprocessed/olives/olives_fundus.pt'

os.remove(meta)
os.remove(pt)

# Confirm deletion
print(f"metadata.json exists: {os.path.exists(meta)}")
print(f"olives_fundus.pt exists: {os.path.exists(pt)}")
print("Ready — now run Cell 8")

metadata.json exists: False
olives_fundus.pt exists: False
Ready — now run Cell 8


In [ ]:
# Check what commit is running and what the matching logic looks like
!git -C /content/dr-dissertation log --oneline -5
!echo "---"
!grep -n "scratch_key\|csv_key\|parts\[0\]\|parts\[2\]\|FUNDUS_TOKEN\|fundus_token\|_normalise" \
    /content/dr-dissertation/src/data/preprocess.py

7521368 (HEAD -> main, origin/main, origin/HEAD) Fix OLIVES path matching using patient/visit/eye key
9adb96d Fix OLIVES CSV column names to match actual dataset
6fcad3d Add missing _extract_fundus_to_scratch helper for OLIVES preprocessing
bc25df7 Update eyepacs_source to local SSD path
d25881c Fix eyepacs_source path to MyDrive root
---
25:FUNDUS_TOKEN = "fundus"
109:        return (trial, parts[2], parts[3], parts[4])
128:    trial = parts[0]
130:        return (trial, parts[1], parts[2], parts[3])
132:        return (trial, parts[2], parts[3], parts[4])
281:                                if FUNDUS_TOKEN in f.lower()
308:    if "TREX" in parts[0]:
311:            "patient_id": parts[2] if len(parts) > 2 else "",
318:        "visit": parts[2] if len(parts) > 2 else "",
382:        if p.is_file() and FUNDUS_TOKEN in p.name.lower()


In [ ]:
!sed -n '90,150p' /content/dr-dissertation/src/data/preprocess.py

    with open(output_dir / METADATA_FILENAME, "w") as f:
        json.dump(metadata, f, indent=2)


def _csv_path_to_key(
    csv_path: str,
) -> tuple[str, str, str, str] | None:
    """
    Extract (trial, patient, visit, eye) from a CSV Path entry.

    CSV format has a leading slash:
      /Prime_FULL/patient/visit/eye/filename.tif → parts[1, 2, 3, 4]
      /TREX DME/arm/patient/visit/eye/filename.tif → parts[1, 3, 4, 5]
    """
    parts = str(csv_path).split("/")
    if len(parts) < 2:
        return None
    trial = parts[1]
    if trial == "Prime_FULL" and len(parts) >= 5:
        return (trial, parts[2], parts[3], parts[4])
    if trial == "TREX DME" and len(parts) >= 6:
        return (trial, parts[3], parts[4], parts[5])
    return None


def _disk_path_to_key(
    rel_path: str,
) -> tuple[str, str, str, str] | None:
    """
    Extract (trial, patient, visit, eye) from a scratch-relative fundus path.

    Disk format has no leading slash:
      Prime_FULL/patient/visit/eye

In [ ]:
!sed -n '270,400p' /content/dr-dissertation/src/data/preprocess.py

                        else "Prime_FULL"
                    )
                    print(f"  Processing {inner_zip_name}...")

                    with olives_zip.open(inner_zip_name) as inner_stream:
                        with zipfile.ZipFile(inner_stream, "r") as inner_zip:

                            # Find all fundus files
                            fundus_files = [
                                f
                                for f in inner_zip.namelist()
                                if FUNDUS_TOKEN in f.lower()
                                and not f.endswith("/")
                            ]
                            print(
                                f"    Found {len(fundus_files)} "
                                f"fundus files in {trial}"
                            )

                            # Extract each fundus file preserving
                            # folder structure
                            for file_path in tqdm(
                        

In [ ]:
import os
from omegaconf import OmegaConf

cfg = OmegaConf.load('/content/dr-dissertation/configs/preprocess.yaml')
print("olives_output:", cfg.olives_output)
print("scratch_dir:", cfg.scratch_dir)

output_path = cfg.olives_output + '/olives_fundus.pt'
print(f"\noutput_path: {output_path}")
print(f"exists: {os.path.exists(output_path)}")

# Also check what's actually in the olives output dir
import os
olives_dir = cfg.olives_output
if os.path.exists(olives_dir):
    for f in os.listdir(olives_dir):
        size = os.path.getsize(f"{olives_dir}/{f}") / 1e6
        print(f"  {f} ({size:.1f} MB)")
else:
    print("olives output dir does not exist")

olives_output: /content/drive/MyDrive/dissertation/preprocessed/olives
scratch_dir: /content/scratch/olives_fundus

output_path: /content/drive/MyDrive/dissertation/preprocessed/olives/olives_fundus.pt
exists: True
  olives_fundus.pt (163.8 MB)
  metadata.json (0.0 MB)


In [ ]:
import os

meta = '/content/drive/MyDrive/dissertation/preprocessed/olives/metadata.json'
pt = '/content/drive/MyDrive/dissertation/preprocessed/olives/olives_fundus.pt'

for path in [meta, pt]:
    if os.path.exists(path):
        os.remove(path)
        print(f"Deleted: {path}")
    else:
        print(f"Already gone: {path}")

# Confirm
print(f"\nmetadata.json exists: {os.path.exists(meta)}")
print(f"olives_fundus.pt exists: {os.path.exists(pt)}")

Deleted: /content/drive/MyDrive/dissertation/preprocessed/olives/metadata.json
Deleted: /content/drive/MyDrive/dissertation/preprocessed/olives/olives_fundus.pt

metadata.json exists: False
olives_fundus.pt exists: False


# Debugging session 2 — clinical xlsx column name bug

**Bug found**: the preprocessing code expected a column named `Path (Trial/Arm/Folder/Visit/Eye/Image Name)` in the clinical xlsx file (matching the biomarker CSV). The clinical xlsx actually uses `File_Path`. This caused a silent column lookup failure.

**Fix**: introduced separate column name constants for the two label files.


In [ ]:
import pandas as pd
from pathlib import Path

PATH_COLUMN = 'Path (Trial/Arm/Folder/Visit/Eye/Image Name)'
FUNDUS_TOKEN = "fundus"

csv_path = '/content/drive/MyDrive/olives_labels/OLIVES_Dataset_Labels/ml_centric_labels/Biomarker_Clinical_Data_Images.csv'
scratch_dir = Path('/content/scratch/olives_fundus')

labels_df = pd.read_csv(csv_path)

def _csv_path_to_key(csv_path):
    parts = str(csv_path).split("/")
    if len(parts) < 2:
        return None
    trial = parts[1]
    if trial == "Prime_FULL" and len(parts) >= 5:
        return (trial, parts[2], parts[3], parts[4])
    if trial == "TREX DME" and len(parts) >= 6:
        return (trial, parts[3], parts[4], parts[5])
    return None

def _disk_path_to_key(rel_path):
    parts = rel_path.split("/")
    if not parts:
        return None
    trial = parts[0]
    if trial == "Prime_FULL" and len(parts) >= 4:
        return (trial, parts[1], parts[2], parts[3])
    if trial == "TREX DME" and len(parts) >= 5:
        return (trial, parts[2], parts[3], parts[4])
    return None

# Build CSV lookup
csv_lookup = {}
for idx, p in enumerate(labels_df[PATH_COLUMN]):
    key = _csv_path_to_key(str(p))
    if key is not None:
        csv_lookup.setdefault(key, []).append(idx)

print(f"CSV unique keys: {len(csv_lookup)}")

# Walk scratch
fundus_files = [
    p for p in scratch_dir.rglob("*")
    if p.is_file() and FUNDUS_TOKEN in p.name.lower()
]
print(f"Fundus files on disk: {len(fundus_files)}")

# Match
matched, unmatched = [], []
for path in fundus_files:
    rel = path.relative_to(scratch_dir).as_posix()
    key = _disk_path_to_key(rel)
    if key in csv_lookup:
        matched.append((rel, key))
    else:
        unmatched.append((rel, key))

print(f"Matched: {len(matched)}")
print(f"Unmatched: {len(unmatched)}")

print("\n=== Sample matched ===")
for rel, key in matched[:3]:
    print(f"  rel: {rel}")
    print(f"  key: {key}")

print("\n=== Sample unmatched ===")
for rel, key in unmatched[:5]:
    print(f"  rel: {rel}")
    print(f"  key: {key}")

# Check if unmatched keys are close to CSV keys
print("\n=== Sample CSV keys (TREX DME) ===")
trex = [k for k in list(csv_lookup.keys()) if k[0] == 'TREX DME'][:3]
for k in trex:
    print(f"  {k}")

print("\n=== Sample CSV keys (Prime_FULL) ===")
prime = [k for k in list(csv_lookup.keys()) if k[0] == 'Prime_FULL'][:3]
for k in prime:
    print(f"  {k}")

CSV unique keys: 192
Fundus files on disk: 4435
Matched: 272
Unmatched: 4163

=== Sample matched ===
  rel: TREX DME/TREX/0213TOS/V19/OS/fundus_OS_V19.tif
  key: ('TREX DME', '0213TOS', 'V19', 'OS')
  rel: TREX DME/TREX/0213TOS/V1/OS/fundus_OS_V1.tif
  key: ('TREX DME', '0213TOS', 'V1', 'OS')
  rel: TREX DME/TREX/0209TOD/V10/OD/fundus_OD_V10.tif
  key: ('TREX DME', '0209TOD', 'V10', 'OD')

=== Sample unmatched ===
  rel: TREX DME/TREX/0213TOS/V14/OD/fundus_OD_V14.tif
  key: ('TREX DME', '0213TOS', 'V14', 'OD')
  rel: TREX DME/TREX/0213TOS/V14/OS/fundus_OS_V14.tif
  key: ('TREX DME', '0213TOS', 'V14', 'OS')
  rel: TREX DME/TREX/0213TOS/V4/OD/fundus_OD_V4.tif
  key: ('TREX DME', '0213TOS', 'V4', 'OD')
  rel: TREX DME/TREX/0213TOS/V4/OS/fundus_OS_V4.tif
  key: ('TREX DME', '0213TOS', 'V4', 'OS')
  rel: TREX DME/TREX/0213TOS/V9/OD/fundus_OD_V9.tif
  key: ('TREX DME', '0213TOS', 'V9', 'OD')

=== Sample CSV keys (TREX DME) ===
  ('TREX DME', '0201GOD', 'V1', 'OD')
  ('TREX DME', '0201GOD', '

In [ ]:
import pandas as pd
from pathlib import Path

# Find OLIVES CSVs - adjust the search root to wherever your OLIVES data lives
search_roots = [
    Path("/content/drive/MyDrive/OLIVES"),
    Path("/content/OLIVES"),
    Path.cwd(),
]

print("=== Searching for CSV files ===")
for root in search_roots:
    if root.exists():
        csvs = list(root.rglob("*.csv"))
        if csvs:
            print(f"\nIn {root}:")
            for c in csvs:
                print(f"  {c}  ({c.stat().st_size / 1024:.1f} KB)")

=== Searching for CSV files ===


In [ ]:
import pandas as pd

clinical_full = "/content/drive/MyDrive/olives_labels/OLIVES_Dataset_Labels/full_labels/Clinical_Data_Images.xlsx"
clinical_ml = "/content/drive/MyDrive/olives_labels/OLIVES_Dataset_Labels/ml_centric_labels/Clinical_Data_Images.xlsx"

for label, path in [("full_labels", clinical_full), ("ml_centric", clinical_ml)]:
    print(f"\n=== {label} Clinical_Data_Images.xlsx ===")
    df = pd.read_excel(path)
    print(f"Shape: {df.shape}")
    print(f"Columns: {list(df.columns)}")
    print(f"\nFirst 3 rows:")
    print(df.head(3).to_string())

    # Find the path column (might be named differently)
    path_cols = [c for c in df.columns if 'path' in c.lower() or 'image' in c.lower() or 'file' in c.lower()]
    print(f"\nPath-like columns: {path_cols}")


=== full_labels Clinical_Data_Images.xlsx ===
Shape: (78185, 5)
Columns: ['File_Path', 'BCVA', 'CST', 'Eye_ID', 'Patient_ID']

First 3 rows:
                                       File_Path  BCVA    CST  Eye_ID  Patient_ID
0  /TREX DME/GILA/0234GOD/V1/OD/TREXW_000027.tif  73.0  517.0      12         234
1  /TREX DME/GILA/0234GOD/V1/OD/TREXW_000000.tif  73.0  517.0      12         234
2  /TREX DME/GILA/0234GOD/V1/OD/TREXW_000001.tif  73.0  517.0      12         234

Path-like columns: ['File_Path']

=== ml_centric Clinical_Data_Images.xlsx ===
Shape: (78185, 5)
Columns: ['File_Path', 'BCVA', 'CST', 'Eye_ID', 'Patient_ID']

First 3 rows:
                                       File_Path  BCVA    CST  Eye_ID  Patient_ID
0  /TREX DME/GILA/0234GOD/V1/OD/TREXW_000027.tif  73.0  517.0      12         234
1  /TREX DME/GILA/0234GOD/V1/OD/TREXW_000000.tif  73.0  517.0      12         234
2  /TREX DME/GILA/0234GOD/V1/OD/TREXW_000001.tif  73.0  517.0      12         234

Path-like columns: ['File_

In [ ]:
for name in ["OCT-DR.xlsx", "OCT-DME.xlsx"]:
    path = f"/content/drive/MyDrive/olives_labels/OLIVES_Dataset_Labels/full_labels/{name}"
    df = pd.read_excel(path)
    print(f"\n=== {name} ===")
    print(f"Shape: {df.shape}")
    print(f"Columns: {list(df.columns)}")
    print(df.head(3).to_string())


=== OCT-DR.xlsx ===
Shape: (41, 20)
Columns: ['Patient Information', 'Unnamed: 1', 'Unnamed: 2', 'Unnamed: 3', 'Unnamed: 4', 'Unnamed: 5', 'Unnamed: 6', 'Diabetes', 'Unnamed: 8', 'Unnamed: 9', 'Unnamed: 10', 'Unnamed: 11', 'Unnamed: 12', 'Unnamed: 13', 'Unnamed: 14', 'Baseline', 'Unnamed: 16', 'Unnamed: 17', 'Unnamed: 18', 'Unnamed: 19']
  Patient Information     Unnamed: 1   Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6            Diabetes                     Unnamed: 8      Unnamed: 9 Unnamed: 10 Unnamed: 11 Unnamed: 12 Unnamed: 13   Unnamed: 14    Baseline Unnamed: 16 Unnamed: 17 Unnamed: 18    Unnamed: 19
0        Patient \nID  Treatment Arm  Study\n Eye        Age     Gender  Ethnicity       Race  Type of\n Diabetes  Number of Years with Diabetes  Baseline HbA1c   W24 HbA1c   W52 HbA1c   W76 HbA1c  W104 HbA1c  BMI (kg/m^2)  ETDRS BCVA         CST   Injection        DRSS  Leakage Index
1              01-001              2           OS         44          M      N H/L     

In [ ]:
import pandas as pd
from pathlib import Path

clin_df = pd.read_excel(
    "/content/drive/MyDrive/olives_labels/OLIVES_Dataset_Labels/full_labels/Clinical_Data_Images.xlsx"
)

def csv_path_to_key(path_str):
    parts = str(path_str).split("/")
    if len(parts) < 2:
        return None
    parts = [p for p in parts if p]  # drop empty from leading /
    trial = parts[0]
    if trial == "Prime_FULL" and len(parts) >= 4:
        return (trial, parts[1], parts[2], parts[3])
    if trial == "TREX DME" and len(parts) >= 5:
        return (trial, parts[2], parts[3], parts[4])
    return None

clin_df['key'] = clin_df['File_Path'].apply(csv_path_to_key)
print(f"Clinical CSV rows: {len(clin_df)}")
print(f"Unique visit-eye keys: {clin_df['key'].nunique()}")
print(f"\nKeys per trial:")
print(clin_df.groupby(clin_df['key'].apply(lambda k: k[0] if k else None))['key'].nunique())

# Compare to biomarker CSV
bio_df = pd.read_csv(
    "/content/drive/MyDrive/olives_labels/OLIVES_Dataset_Labels/ml_centric_labels/Biomarker_Clinical_Data_Images.csv"
)
PATH_COL = 'Path (Trial/Arm/Folder/Visit/Eye/Image Name)'
bio_df['key'] = bio_df[PATH_COL].apply(csv_path_to_key)
print(f"\nBiomarker CSV unique keys: {bio_df['key'].nunique()}")
print(f"Clinical-only keys (BCVA/CST without biomarkers): {clin_df['key'].nunique() - len(set(clin_df['key']) & set(bio_df['key']))}")

Clinical CSV rows: 78185
Unique visit-eye keys: 1594

Keys per trial:
key
Prime_FULL    660
TREX DME      934
Name: key, dtype: int64

Biomarker CSV unique keys: 192
Clinical-only keys (BCVA/CST without biomarkers): 1404


In [ ]:
import pandas as pd

clin_df = pd.read_excel(
    "/content/drive/MyDrive/olives_labels/OLIVES_Dataset_Labels/full_labels/Clinical_Data_Images.xlsx"
)

def csv_path_to_key(path_str):
    parts = [p for p in str(path_str).split("/") if p]
    if len(parts) < 4:
        return None
    trial = parts[0]
    if trial == "Prime_FULL" and len(parts) >= 4:
        return (trial, parts[1], parts[2], parts[3])
    if trial == "TREX DME" and len(parts) >= 5:
        return (trial, parts[2], parts[3], parts[4])
    return None

clin_df['key'] = clin_df['File_Path'].apply(csv_path_to_key)

# Now the consistency check
inconsistent = clin_df.dropna(subset=['key']).groupby('key').agg({
    'BCVA': 'nunique',
    'CST': 'nunique'
})
print(f"Total visit-eye keys: {len(inconsistent)}")
print(f"Keys with inconsistent BCVA across OCT scans: {(inconsistent['BCVA'] > 1).sum()}")
print(f"Keys with inconsistent CST across OCT scans: {(inconsistent['CST'] > 1).sum()}")

# If any are inconsistent, show a sample so we can see what's going on
if (inconsistent['BCVA'] > 1).any() or (inconsistent['CST'] > 1).any():
    bad_keys = inconsistent[(inconsistent['BCVA'] > 1) | (inconsistent['CST'] > 1)].index[:3]
    print("\nSample inconsistent keys:")
    for k in bad_keys:
        sample = clin_df[clin_df['key'] == k][['File_Path', 'BCVA', 'CST']].head(5)
        print(f"\nKey: {k}")
        print(sample.to_string())

Total visit-eye keys: 1594
Keys with inconsistent BCVA across OCT scans: 0
Keys with inconsistent CST across OCT scans: 0


In [ ]:
import os

# First, make sure Drive is mounted
if not os.path.exists('/content/drive/MyDrive'):
    from google.colab import drive
    drive.mount('/content/drive')

# Search for your repo by looking for the .git folder
!find /content/drive/MyDrive -maxdepth 4 -name ".git" -type d 2>/dev/null

Mounted at /content/drive


In [ ]:
import os

REPO_DIR = "/content/dr-dissertation"
GITHUB_USER = "savita10"
REPO_NAME = "dr-dissertation"
from google.colab import userdata
GITHUB_TOKEN = userdata.get('GITHUB_TOKEN')  # Reads from Colab Secrets — token never appears in code

if not os.path.exists(REPO_DIR):
    !git clone https://{GITHUB_USER}:{GITHUB_TOKEN}@github.com/{GITHUB_USER}/{REPO_NAME}.git {REPO_DIR}
    print("Cloned fresh")
else:
    !git -C {REPO_DIR} pull
    print("Pulled latest")

Cloning into '/content/dr-dissertation'...
remote: Enumerating objects: 100, done.
remote: Counting objects: 100% (100/100), done.
remote: Compressing objects: 100% (63/63), done.
remote: Total 100 (delta 46), reused 84 (delta 30), pack-reused 0 (from 0)
Receiving objects: 100% (100/100), 31.38 KiB | 10.46 MiB/s, done.
Resolving deltas: 100% (46/46), done.
Cloned fresh


In [ ]:
# Remove stale outputs from the previous 272-match run
!rm -f /content/drive/MyDrive/dissertation/preprocessed/olives/olives_fundus.pt
!rm -f /content/drive/MyDrive/dissertation/preprocessed/olives/metadata.json

# Verify they're gone
!ls -la /content/drive/MyDrive/dissertation/preprocessed/olives/

# Run the new preprocessing
!cd /content/dr-dissertation && python -m src.data.preprocess --dataset olives

total 0

OLIVES preprocessing

[1/6 Extracting fundus images to /content/scratch/olives_fundus]
  Opening outer zip: /content/drive/MyDrive/olive.zip
  Opening OLIVES.zip as stream...
  Processing OLIVES/TREX_DME.zip...
    Found 1849 fundus files in TREX_DME
Extracting TREX_DME: 100% 1849/1849 [02:06<00:00, 14.63it/s]
  Processing OLIVES/Prime_FULL.zip...
    Found 2586 fundus files in Prime_FULL
Extracting Prime_FULL: 100% 2586/2586 [01:43<00:00, 24.99it/s]
extraction complete in 2793.2s

[2/6 Loading biomarker labels from /content/drive/MyDrive/olives_labels/OLIVES_Dataset_Labels/ml_centric_labels/Biomarker_Clinical_Data_Images.csv]
Biomarker lookup: 192 keys (loaded in 0.9s)

[3/6 Loading clinical labels from /content/drive/MyDrive/olives_labels/OLIVES_Dataset_Labels/full_labels/Clinical_Data_Images.xlsx]

[ERROR] olives preprocessing failed: "Expected column 'Path (Trial/Arm/Folder/Visit/Eye/Image Name)' in /content/drive/MyDrive/olives_labels/OLIVES_Dataset_Labels/full_labels/Cli

In [ ]:
from google.colab import userdata
TOKEN = userdata.get('GITHUB_TOKEN')  # Reads from Colab Secrets

!git -C /content/dr-dissertation pull \
    https://savita10:{TOKEN}@github.com/savita10/dr-dissertation.git
print("Pulled latest")


remote: Enumerating objects: 9, done.
remote: Counting objects: 100% (9/9), done.
remote: Compressing objects: 100% (2/2), done.
remote: Total 5 (delta 3), reused 5 (delta 3), pack-reused 0 (from 0)
Unpacking objects: 100% (5/5), 832 bytes | 416.00 KiB/s, done.
From https://github.com/savita10/dr-dissertation
 * branch            HEAD       -> FETCH_HEAD
Updating 31432d9..27789ae
Fast-forward
 src/data/preprocess.py | 14 +++++++++++---
 1 file changed, 11 insertions(+), 3 deletions(-)
Pulled latest


In [ ]:
!cd /content/dr-dissertation && python -m src.data.preprocess --dataset olives


OLIVES preprocessing

[1/6 Extracting fundus images to /content/scratch/olives_fundus]
  Opening outer zip: /content/drive/MyDrive/olive.zip
  Opening OLIVES.zip as stream...
  Processing OLIVES/TREX_DME.zip...
    Found 1849 fundus files in TREX_DME
Extracting TREX_DME: 100% 1849/1849 [00:00<00:00, 29015.52it/s]
  Processing OLIVES/Prime_FULL.zip...
    Found 2586 fundus files in Prime_FULL
Extracting Prime_FULL: 100% 2586/2586 [00:00<00:00, 51104.50it/s]
extraction complete in 1182.4s

[2/6 Loading biomarker labels from /content/drive/MyDrive/olives_labels/OLIVES_Dataset_Labels/ml_centric_labels/Biomarker_Clinical_Data_Images.csv]
Biomarker lookup: 192 keys (loaded in 0.1s)

[3/6 Loading clinical labels from /content/drive/MyDrive/olives_labels/OLIVES_Dataset_Labels/full_labels/Clinical_Data_Images.xlsx]
Clinical lookup: 1594 keys (loaded in 4.7s)

[4/6 Walking scratch dir for fundus files]
found 4435 fundus files on disk

[5/6 Preprocessing all fundus images with three-tier label a

# Debugging session 3 — biomarker deduplication losing positives

**Bug found**: biomarker annotations are per-OCT-scan (one row per scan), but our pipeline needed visit-eye level labels. The original code used `drop_duplicates(keep='first')`, which silently lost 60-70% of biomarker positives — a biomarker might be visible on scan 27 but absent on scans 1-26, and `keep='first'` would record the visit-eye as negative.

**Fix**: replaced with `groupby('_key')[biomarker_cols].max()` — an "any-positive" aggregation rule that correctly captures "positive on at least one scan".

This is a clinically defensible decision rule (a clinician would summarise pathology as present at the visit level if it's visible anywhere in the OCT volume) and is worth flagging in the dissertation methods chapter as a deliberate methodological choice.


In [ ]:
import torch

data = torch.load(
    "/content/drive/MyDrive/dissertation/preprocessed/olives/olives_fundus.pt",
    map_location="cpu",
    weights_only=False,
)

print("=== Tensor shapes ===")
for key, val in data.items():
    if torch.is_tensor(val):
        print(f"  {key:20s} {tuple(val.shape)}  {val.dtype}")
    elif isinstance(val, list):
        print(f"  {key:20s} list of {len(val)} items")

print("\n=== Mask consistency ===")
print(f"  has_clinical sum:     {data['has_clinical'].sum().item()}  (expect 2254)")
print(f"  has_biomarkers sum:   {data['has_biomarkers'].sum().item()}  (expect 272)")

print("\n=== NaN distribution ===")
print(f"  bcva NaN count:       {torch.isnan(data['bcva']).sum().item()}  (expect 4435 - 2254 = 2181)")
print(f"  cst NaN count:        {torch.isnan(data['cst']).sum().item()}  (expect 2181)")
print(f"  biomarkers NaN cells: {torch.isnan(data['biomarkers']).sum().item()}  (expect (4435-272)*16 + small intra-row NaNs)")

print("\n=== Patient ID coverage ===")
valid_patient_ids = data['patient_ids'][data['patient_ids'] >= 0]
print(f"  Unique patient IDs:   {valid_patient_ids.unique().numel()}")
print(f"  Missing patient IDs:  {(data['patient_ids'] == -1).sum().item()}")

print("\n=== DRIL label distribution ===")
labels_with_biomarkers = data['labels'][data['has_biomarkers']]
unique, counts = torch.unique(labels_with_biomarkers, return_counts=True)
for u, c in zip(unique.tolist(), counts.tolist()):
    print(f"  DRIL={u}: {c} samples")

print("\n=== Sample tier-0 entry ===")
tier_0_mask = ~data['has_clinical'] & ~data['has_biomarkers']
if tier_0_mask.any():
    idx = tier_0_mask.nonzero()[0].item()
    print(f"  index {idx}: key={data['keys'][idx]}, bcva={data['bcva'][idx]}, has_clinical={data['has_clinical'][idx].item()}, has_biomarkers={data['has_biomarkers'][idx].item()}")
print(f"  Total tier-0 samples: {tier_0_mask.sum().item()} (expect 4435 - 2254 = 2181)")

=== Tensor shapes ===
  images               (4435, 3, 224, 224)  torch.float32
  labels               (4435,)  torch.int64
  metadata             list of 4435 items
  biomarkers           (4435, 16)  torch.float32
  bcva                 (4435,)  torch.float32
  cst                  (4435,)  torch.float32
  has_clinical         (4435,)  torch.bool
  has_biomarkers       (4435,)  torch.bool
  patient_ids          (4435,)  torch.int64
  eye_ids              (4435,)  torch.int64
  keys                 list of 4435 items

=== Mask consistency ===
  has_clinical sum:     2254  (expect 2254)
  has_biomarkers sum:   272  (expect 272)

=== NaN distribution ===
  bcva NaN count:       2182  (expect 4435 - 2254 = 2181)
  cst NaN count:        2182  (expect 2181)
  biomarkers NaN cells: 66615  (expect (4435-272)*16 + small intra-row NaNs)

=== Patient ID coverage ===
  Unique patient IDs:   87
  Missing patient IDs:  2181

=== DRIL label distribution ===
  DRIL=0: 272 samples

=== Sample tier-0 e

In [ ]:
import torch

data = torch.load(
    "/content/drive/MyDrive/dissertation/preprocessed/olives/olives_fundus.pt",
    map_location="cpu",
    weights_only=False,
)

# Find samples flagged as having clinical labels but with NaN BCVA or CST
clinical_with_nan_bcva = data['has_clinical'] & torch.isnan(data['bcva'])
clinical_with_nan_cst = data['has_clinical'] & torch.isnan(data['cst'])

print(f"has_clinical=True but BCVA is NaN: {clinical_with_nan_bcva.sum().item()}")
print(f"has_clinical=True but CST is NaN:  {clinical_with_nan_cst.sum().item()}")

# Show the affected keys
if clinical_with_nan_bcva.any():
    print("\nAffected samples (has_clinical=True but BCVA=NaN):")
    indices = clinical_with_nan_bcva.nonzero().squeeze(-1).tolist()
    for idx in indices[:5]:
        print(f"  idx={idx}, key={data['keys'][idx]}, bcva={data['bcva'][idx].item()}, cst={data['cst'][idx].item()}, patient_id={data['patient_ids'][idx].item()}")

# Verify tier accounting
print("\n=== Tier verification ===")
tier_0 = ~data['has_clinical'] & ~data['has_biomarkers']
tier_1 = data['has_clinical'] & ~data['has_biomarkers']
tier_2 = data['has_clinical'] & data['has_biomarkers']
biomarker_only = ~data['has_clinical'] & data['has_biomarkers']

print(f"Tier 0 (no labels):           {tier_0.sum().item()}")
print(f"Tier 1 (clinical only):       {tier_1.sum().item()}")
print(f"Tier 2 (clinical + biomarker): {tier_2.sum().item()}")
print(f"Biomarker only (anomalous):   {biomarker_only.sum().item()}")
print(f"Total:                        {tier_0.sum().item() + tier_1.sum().item() + tier_2.sum().item() + biomarker_only.sum().item()}")

has_clinical=True but BCVA is NaN: 1
has_clinical=True but CST is NaN:  1

Affected samples (has_clinical=True but BCVA=NaN):
  idx=942, key=('TREX DME', '0247GOD', 'V3', 'OD'), bcva=nan, cst=nan, patient_id=247

=== Tier verification ===
Tier 0 (no labels):           2177
Tier 1 (clinical only):       1986
Tier 2 (clinical + biomarker): 268
Biomarker only (anomalous):   4
Total:                        4435


In [ ]:
import pandas as pd

clin_df = pd.read_excel("/content/drive/MyDrive/olives_labels/OLIVES_Dataset_Labels/full_labels/Clinical_Data_Images.xlsx")

# How many unique visit-eye keys have ANY NaN BCVA?
nan_rows = clin_df[clin_df['BCVA'].isna()]
print(f"NaN BCVA rows: {len(nan_rows)}")

# Build keys for these NaN rows
def csv_path_to_key(path_str):
    parts = [p for p in str(path_str).split("/") if p]
    if len(parts) < 4:
        return None
    trial = parts[0]
    if trial == "Prime_FULL" and len(parts) >= 4:
        return (trial, parts[1], parts[2], parts[3])
    if trial == "TREX DME" and len(parts) >= 5:
        return (trial, parts[2], parts[3], parts[4])
    return None

nan_rows = nan_rows.copy()
nan_rows['key'] = nan_rows['File_Path'].apply(csv_path_to_key)
nan_keys = nan_rows['key'].dropna().unique()
print(f"Unique visit-eye keys with NaN BCVA: {len(nan_keys)}")
print(f"\nThese keys:")
for k in nan_keys:
    print(f"  {k}")

NaN BCVA rows: 49
Unique visit-eye keys with NaN BCVA: 1

These keys:
  ('TREX DME', '0247GOD', 'V3', 'OD')


In [ ]:
import torch

data = torch.load(
    "/content/drive/MyDrive/dissertation/preprocessed/olives/olives_fundus.pt",
    weights_only=False,
)

biomarker_names = [
    'Atrophy/thinning', 'EZ Disruption', 'DRIL', 'IR hemorrhages',
    'IR HRF', 'Vitreous (partial)', 'Vitreous (full)', 'Preretinal',
    'Vitreous debris', 'VMT', 'DRT/ME', 'IRF', 'SRF', 'RPE Disruption',
    'PED (serous)', 'SHRM'
]

biomarkers = data['biomarkers'][data['has_biomarkers']]
print(f"Per-biomarker positive counts (out of {len(biomarkers)} tier-2 samples):\n")
print(f"  {'Biomarker':<25s} {'Pos':>5s} {'Valid':>6s} {'%':>6s}")
print(f"  {'-'*25} {'-'*5} {'-'*6} {'-'*6}")
for i, name in enumerate(biomarker_names):
    col = biomarkers[:, i]
    valid = col[~torch.isnan(col)]
    pos = (valid > 0).sum().item()
    pct = pos / len(valid) * 100 if len(valid) > 0 else 0
    print(f"  {name:<25s} {pos:>5d} {len(valid):>6d} {pct:>5.1f}%")

Per-biomarker positive counts (out of 272 tier-2 samples):

  Biomarker                   Pos  Valid      %
  ------------------------- ----- ------ ------
  Atrophy/thinning              4    272   1.5%
  EZ Disruption                10    272   3.7%
  DRIL                          0    272   0.0%
  IR hemorrhages               21    272   7.7%
  IR HRF                      124    272  45.6%
  Vitreous (partial)           79    270  29.3%
  Vitreous (full)             151    270  55.9%
  Preretinal                   13    271   4.8%
  Vitreous debris             122    270  45.2%
  VMT                           0    272   0.0%
  DRT/ME                       39    272  14.3%
  IRF                          57    272  21.0%
  SRF                           0    272   0.0%
  RPE Disruption                0    272   0.0%
  PED (serous)                  0    272   0.0%
  SHRM                          0    272   0.0%


In [ ]:
import pandas as pd
bio = pd.read_csv("/content/drive/MyDrive/olives_labels/OLIVES_Dataset_Labels/ml_centric_labels/Biomarker_Clinical_Data_Images.csv")
print("Original biomarker CSV — value counts before deduplication:")
for col in ['DRIL', 'VMT', 'Fluid (SRF)', 'Disruption of RPE', 'PED (serous)', 'SHRM']:
    print(f"\n  {col}:")
    print(bio[col].value_counts(dropna=False))

Original biomarker CSV — value counts before deduplication:

  DRIL:
DRIL
0    9376
1      32
Name: count, dtype: int64

  VMT:
VMT
0    9398
1      10
Name: count, dtype: int64

  Fluid (SRF):
Fluid (SRF)
0    9175
1     233
Name: count, dtype: int64

  Disruption of RPE:
Disruption of RPE
0    9398
1      10
Name: count, dtype: int64

  PED (serous):
PED (serous)
0    9398
1      10
Name: count, dtype: int64

  SHRM:
SHRM
0    9332
1      76
Name: count, dtype: int64


In [ ]:
import pandas as pd

bio = pd.read_csv("/content/drive/MyDrive/olives_labels/OLIVES_Dataset_Labels/ml_centric_labels/Biomarker_Clinical_Data_Images.csv")

PATH_COL = 'Path (Trial/Arm/Folder/Visit/Eye/Image Name)'

def csv_path_to_key(path_str):
    parts = [p for p in str(path_str).split("/") if p]
    if len(parts) < 4:
        return None
    trial = parts[0]
    if trial == "Prime_FULL" and len(parts) >= 4:
        return (trial, parts[1], parts[2], parts[3])
    if trial == "TREX DME" and len(parts) >= 5:
        return (trial, parts[2], parts[3], parts[4])
    return None

bio['key'] = bio[PATH_COL].apply(csv_path_to_key)
bio = bio.dropna(subset=['key'])

biomarker_cols = [
    'Atrophy / thinning of retinal layers', 'Disruption of EZ', 'DRIL',
    'IR hemorrhages', 'IR HRF', 'Partially attached vitreous face',
    'Fully attached vitreous face', 'Preretinal tissue/hemorrhage',
    'Vitreous debris', 'VMT', 'DRT/ME', 'Fluid (IRF)', 'Fluid (SRF)',
    'Disruption of RPE', 'PED (serous)', 'SHRM'
]

# How many keys have intra-visit variation per biomarker?
print(f"Intra-visit-eye variation per biomarker (out of 192 keys):\n")
print(f"  {'Biomarker':<35s} {'Varies':>7s} {'Max=1':>7s} {'First=1':>8s}")
print(f"  {'-'*35} {'-'*7} {'-'*7} {'-'*8}")
for col in biomarker_cols:
    grouped = bio.groupby('key')[col]
    n_varies = (grouped.nunique() > 1).sum()
    n_max_1 = (grouped.max() > 0).sum()
    n_first_1 = (grouped.first() > 0).sum()
    print(f"  {col:<35s} {n_varies:>7d} {n_max_1:>7d} {n_first_1:>8d}")

Intra-visit-eye variation per biomarker (out of 192 keys):

  Biomarker                            Varies   Max=1  First=1
  ----------------------------------- ------- ------- --------
  Atrophy / thinning of retinal layers      16      17        4
  Disruption of EZ                         68      68        7
  DRIL                                      5       5        0
  IR hemorrhages                           67      67       13
  IR HRF                                  158     192       90
  Partially attached vitreous face         93     110       58
  Fully attached vitreous face             62     134      111
  Preretinal tissue/hemorrhage             43      48       12
  Vitreous debris                         162     164       73
  VMT                                       3       3        0
  DRT/ME                                  104     114       31
  Fluid (IRF)                             162     174       44
  Fluid (SRF)                              29      29    

In [ ]:
# Pull the fix
!git -C /content/dr-dissertation pull

# Delete the stale .pt and metadata so the early-exit guard doesn't skip the run
!rm -f /content/drive/MyDrive/dissertation/preprocessed/olives/olives_fundus.pt
!rm -f /content/drive/MyDrive/dissertation/preprocessed/olives/metadata.json

# Re-run preprocessing
!cd /content/dr-dissertation && python -m src.data.preprocess --dataset olives

remote: Enumerating objects: 9, done.
remote: Counting objects: 100% (9/9), done.
remote: Compressing objects: 100% (2/2), done.
remote: Total 5 (delta 3), reused 5 (delta 3), pack-reused 0 (from 0)
Unpacking objects: 100% (5/5), 988 bytes | 329.00 KiB/s, done.
From https://github.com/savita10/dr-dissertation
   31432d9..7e81795  main       -> origin/main
Updating 27789ae..7e81795
Fast-forward
 src/data/preprocess.py | 19 ++++++++++++++++---
 1 file changed, 16 insertions(+), 3 deletions(-)

OLIVES preprocessing

[1/6 Extracting fundus images to /content/scratch/olives_fundus]
  Opening outer zip: /content/drive/MyDrive/olive.zip
  Opening OLIVES.zip as stream...
  Processing OLIVES/TREX_DME.zip...
    Found 1849 fundus files in TREX_DME
Extracting TREX_DME: 100% 1849/1849 [00:00<00:00, 48281.52it/s]
  Processing OLIVES/Prime_FULL.zip...
    Found 2586 fundus files in Prime_FULL
Extracting Prime_FULL: 100% 2586/2586 [00:00<00:00, 34246.57it/s]
extraction complete in 1178.2s

[2/6 Loadi

In [ ]:
import torch

data = torch.load(
    "/content/drive/MyDrive/dissertation/preprocessed/olives/olives_fundus.pt",
    weights_only=False,
)

biomarker_names = [
    'Atrophy/thinning', 'EZ Disruption', 'DRIL', 'IR hemorrhages',
    'IR HRF', 'Vitreous (partial)', 'Vitreous (full)', 'Preretinal',
    'Vitreous debris', 'VMT', 'DRT/ME', 'IRF', 'SRF', 'RPE Disruption',
    'PED (serous)', 'SHRM'
]

biomarkers = data['biomarkers'][data['has_biomarkers']]
print(f"Per-biomarker positive counts (out of {len(biomarkers)} tier-2 samples):\n")
print(f"  {'Biomarker':<25s} {'Pos':>5s} {'Valid':>6s} {'%':>6s}")
print(f"  {'-'*25} {'-'*5} {'-'*6} {'-'*6}")
for i, name in enumerate(biomarker_names):
    col = biomarkers[:, i]
    valid = col[~torch.isnan(col)]
    pos = (valid > 0).sum().item()
    pct = pos / len(valid) * 100 if len(valid) > 0 else 0
    print(f"  {name:<25s} {pos:>5d} {len(valid):>6d} {pct:>5.1f}%")

Per-biomarker positive counts (out of 272 tier-2 samples):

  Biomarker                   Pos  Valid      %
  ------------------------- ----- ------ ------
  Atrophy/thinning             18    272   6.6%
  EZ Disruption                83    272  30.5%
  DRIL                          9    272   3.3%
  IR hemorrhages               94    272  34.6%
  IR HRF                      272    272 100.0%
  Vitreous (partial)          156    272  57.4%
  Vitreous (full)             180    272  66.2%
  Preretinal                   66    272  24.3%
  Vitreous debris             239    272  87.9%
  VMT                           4    272   1.5%
  DRT/ME                      140    272  51.5%
  IRF                         239    272  87.9%
  SRF                          32    272  11.8%
  RPE Disruption                8    272   2.9%
  PED (serous)                  2    272   0.7%
  SHRM                         21    272   7.7%


# Final verification — post-fix dataset state

Cells below confirm the preprocessed OLIVES dataset matches the expected cohort statistics from Prabhushankar et al., NeurIPS 2022 (96 patients, 204 patient-eyes, ~15.4 visits per eye), and check the three-tier label structure is correctly populated.


In [ ]:
import torch
from pathlib import Path
import json

eyepacs_dir = Path("/content/drive/MyDrive/dissertation/preprocessed/eyepacs")

# 1. Check what's there
print("=== Files in EyePACS preprocessed dir ===")
shards = sorted(eyepacs_dir.glob("eyepacs_shard_*.pt"))
for s in shards:
    print(f"  {s.name}  ({s.stat().st_size / 1024 / 1024:.1f} MB)")
print(f"\nTotal shards: {len(shards)}")

# 2. Read metadata
print("\n=== EyePACS metadata.json ===")
with open(eyepacs_dir / "metadata.json") as f:
    meta = json.load(f)
print(json.dumps(meta, indent=2))

# 3. Spot-check one shard
print("\n=== Sample shard contents (first shard) ===")
if shards:
    shard = torch.load(shards[0], map_location="cpu", weights_only=False)
    for key, val in shard.items():
        if torch.is_tensor(val):
            print(f"  {key:15s} shape={tuple(val.shape)}, dtype={val.dtype}")
        elif isinstance(val, list):
            print(f"  {key:15s} list of {len(val)} items, first: {val[0] if val else 'empty'}")

    print(f"\n  Image stats: min={shard['images'].min().item():.3f}, "
          f"max={shard['images'].max().item():.3f}, "
          f"mean={shard['images'].mean().item():.3f}")
    print(f"  Labels in this shard: {torch.unique(shard['labels'], return_counts=True)}")

# 4. Aggregate label distribution across all shards
print("\n=== Aggregate label distribution across all shards ===")
all_labels = []
for s in shards:
    shard = torch.load(s, map_location="cpu", weights_only=False)
    all_labels.append(shard["labels"])
all_labels = torch.cat(all_labels)
unique, counts = torch.unique(all_labels, return_counts=True)
total = len(all_labels)
print(f"Total images: {total}")
print(f"\n  {'DR Grade':<12s} {'Count':>8s} {'%':>6s}")
print(f"  {'-'*12} {'-'*8} {'-'*6}")
for u, c in zip(unique.tolist(), counts.tolist()):
    grade_name = ['No DR', 'Mild', 'Moderate', 'Severe', 'Proliferative'][u] if u < 5 else f"Class {u}"
    print(f"  {grade_name:<12s} {c:>8d} {c/total*100:>5.1f}%")

# 5. Cross-dataset compatibility check
print("\n=== EyePACS vs OLIVES compatibility ===")
olives_data = torch.load(
    "/content/drive/MyDrive/dissertation/preprocessed/olives/olives_fundus.pt",
    weights_only=False,
)
eyepacs_img = shard['images'][0]
olives_img = olives_data['images'][0]
print(f"  EyePACS image shape: {tuple(eyepacs_img.shape)}, dtype: {eyepacs_img.dtype}")
print(f"  OLIVES image shape:  {tuple(olives_img.shape)}, dtype: {olives_img.dtype}")
print(f"  EyePACS image stats: min={eyepacs_img.min():.3f}, max={eyepacs_img.max():.3f}, mean={eyepacs_img.mean():.3f}")
print(f"  OLIVES image stats:  min={olives_img.min():.3f}, max={olives_img.max():.3f}, mean={olives_img.mean():.3f}")
print(f"  Compatible for shared encoder: {eyepacs_img.shape == olives_img.shape and eyepacs_img.dtype == olives_img.dtype}")

=== Files in EyePACS preprocessed dir ===
  eyepacs_shard_000.pt  (2871.2 MB)
  eyepacs_shard_001.pt  (2871.2 MB)
  eyepacs_shard_002.pt  (2871.2 MB)
  eyepacs_shard_003.pt  (2871.2 MB)
  eyepacs_shard_004.pt  (2871.2 MB)
  eyepacs_shard_005.pt  (2871.2 MB)
  eyepacs_shard_006.pt  (2871.2 MB)
  eyepacs_shard_007.pt  (62.0 MB)

Total shards: 8

=== EyePACS metadata.json ===
{
  "total_images": 35108,
  "num_shards": 8,
  "image_size": [
    224,
    224
  ],
  "normalisation": {
    "mean": [
      0.485,
      0.456,
      0.406
    ],
    "std": [
      0.229,
      0.224,
      0.225
    ]
  },
  "class_distribution": {
    "0": 25802,
    "1": 2438,
    "2": 5288,
    "3": 872,
    "4": 708
  },
  "preprocessing_date": "2026-04-26T11:28:38.126998+00:00"
}

=== Sample shard contents (first shard) ===
  images          shape=(5000, 3, 224, 224), dtype=torch.float32
  labels          shape=(5000,), dtype=torch.int64
  filenames       list of 5000 items, first: eyepacs_000000

  Image s

In [ ]:
import torch
from pathlib import Path

# Load OLIVES
olives_data = torch.load(
    "/content/drive/MyDrive/dissertation/preprocessed/olives/olives_fundus.pt",
    weights_only=False,
)
olives_imgs = olives_data['images']

# Load 200 EyePACS images (first shard, first 200)
eyepacs_dir = Path("/content/drive/MyDrive/dissertation/preprocessed/eyepacs")
shard_0 = torch.load(eyepacs_dir / "eyepacs_shard_000.pt", weights_only=False)
eyepacs_imgs = shard_0['images'][:200]

# Sample 200 OLIVES images
n_olives = min(200, len(olives_imgs))
olives_sample = olives_imgs[:n_olives]

# Per-channel stats (after normalisation, so should be roughly mean=0 if both
# datasets matched ImageNet statistics; deviation indicates domain shift)
print(f"=== Per-channel statistics (after ImageNet normalisation) ===")
print(f"  {'Dataset':<10s} {'N':>5s} {'Ch':>4s} {'Mean':>8s} {'Std':>8s} {'Min':>8s} {'Max':>8s}")
print(f"  {'-'*10} {'-'*5} {'-'*4} {'-'*8} {'-'*8} {'-'*8} {'-'*8}")

for name, imgs in [("EyePACS", eyepacs_imgs), ("OLIVES", olives_sample)]:
    for c, channel in enumerate(['R', 'G', 'B']):
        ch_data = imgs[:, c, :, :]
        print(
            f"  {name:<10s} {len(imgs):>5d} {channel:>4s} "
            f"{ch_data.mean().item():>+8.3f} {ch_data.std().item():>8.3f} "
            f"{ch_data.min().item():>+8.3f} {ch_data.max().item():>+8.3f}"
        )

# Aggregate (across all channels) for readability
print(f"\n=== Aggregate (all channels) ===")
print(f"  {'Dataset':<10s} {'Mean':>8s} {'Std':>8s}")
print(f"  {'-'*10} {'-'*8} {'-'*8}")
for name, imgs in [("EyePACS (n=200)", eyepacs_imgs), ("OLIVES (n=200)", olives_sample)]:
    print(f"  {name:<10s} {imgs.mean().item():>+8.3f} {imgs.std().item():>8.3f}")

# Per-dataset proportion of pixels near zero (i.e. background/black border)
# After ImageNet normalisation, raw black (0,0,0) maps to roughly (-2.12, -2.04, -1.80)
print(f"\n=== Black-border / dark-region proportion ===")
print(f"  Pixels with R<-1.5 (i.e. raw R < ~0.06):")
for name, imgs in [("EyePACS", eyepacs_imgs), ("OLIVES", olives_sample)]:
    pct_dark = (imgs[:, 0, :, :] < -1.5).float().mean().item() * 100
    print(f"    {name:<10s} {pct_dark:>6.2f}% of pixels")

=== Per-channel statistics (after ImageNet normalisation) ===
  Dataset        N   Ch     Mean      Std      Min      Max
  ---------- ----- ---- -------- -------- -------- --------
  EyePACS      200    R   -0.250    1.226   -2.118   +2.249
  EyePACS      200    G   -0.697    0.922   -2.036   +2.429
  EyePACS      200    B   -0.869    0.762   -1.804   +2.640
  OLIVES       200    R   -0.093    0.746   -2.118   +2.249
  OLIVES       200    G   +0.034    0.762   -2.036   +2.429
  OLIVES       200    B   +0.256    0.759   -1.804   +2.640

=== Aggregate (all channels) ===
  Dataset        Mean      Std
  ---------- -------- --------
  EyePACS (n=200)   -0.605    1.023
  OLIVES (n=200)   +0.065    0.769

=== Black-border / dark-region proportion ===
  Pixels with R<-1.5 (i.e. raw R < ~0.06):
    EyePACS     20.15% of pixels
    OLIVES       4.62% of pixels


In [ ]:
import torch
from collections import Counter

data = torch.load(
    "/content/drive/MyDrive/dissertation/preprocessed/olives/olives_fundus.pt",
    weights_only=False,
)

# How many unique visit-eyes do we have in our 4,435 fundus dataset?
unique_keys = set(data['keys'])
print(f"Unique visit-eye keys among {len(data['keys'])} fundus files: {len(unique_keys)}")

# Distribution: how many fundus files per visit-eye?
key_counts = Counter(data['keys'])
files_per_key = Counter(key_counts.values())
print(f"\nFundus files per visit-eye:")
for n_files, n_keys in sorted(files_per_key.items()):
    print(f"  {n_files} file(s): {n_keys} visit-eyes")

# Sample a visit-eye with many files to see what the duplication looks like
most_common = max(key_counts.items(), key=lambda x: x[1])
print(f"\nVisit-eye with most fundus files: {most_common[0]} ({most_common[1]} files)")
matching_indices = [i for i, k in enumerate(data['keys']) if k == most_common[0]]
print(f"Their metadata entries:")
for idx in matching_indices[:5]:
    print(f"  idx={idx}: {data['metadata'][idx]}")

Unique visit-eye keys among 4435 fundus files: 3142

Fundus files per visit-eye:
  1 file(s): 1849 visit-eyes
  2 file(s): 1293 visit-eyes

Visit-eye with most fundus files: ('Prime_FULL', '01-012', 'W8', 'OD') (2 files)
Their metadata entries:
  idx=1849: {'trial': 'Prime_FULL', 'patient_id': '01-012', 'visit': 'W8', 'eye': 'OD'}
  idx=1850: {'trial': 'Prime_FULL', 'patient_id': '01-012', 'visit': 'W8', 'eye': 'OD'}


In [ ]:
# Remount Drive
from google.colab import drive
drive.mount('/content/drive', force_remount=True)

# Verify the file exists
import os
path = "/content/drive/MyDrive/dissertation/preprocessed/olives/olives_fundus.pt"
print(f"File exists: {os.path.exists(path)}")
if os.path.exists(path):
    print(f"Size: {os.path.getsize(path) / 1024 / 1024:.1f} MB")

Mounted at /content/drive
File exists: True
Size: 1804.8 MB


In [ ]:
import torch
from collections import Counter

data = torch.load(
    "/content/drive/MyDrive/dissertation/preprocessed/olives/olives_fundus.pt",
    weights_only=False,
)

# Per-trial breakdown
keys = data['keys']
trex_keys = [k for k in keys if k[0] == 'TREX DME']
prime_keys = [k for k in keys if k[0] == 'Prime_FULL']

print(f"TREX DME fundus files: {len(trex_keys)}")
print(f"  Unique visit-eyes:   {len(set(trex_keys))}")
print(f"  Unique patients:     {len(set(k[1] for k in trex_keys))}")
print(f"  Unique patient-eyes: {len(set((k[1], k[3]) for k in trex_keys))}")
print(f"  Unique visits:       {len(set(k[2] for k in trex_keys))}")

print(f"\nPrime_FULL fundus files: {len(prime_keys)}")
print(f"  Unique visit-eyes:   {len(set(prime_keys))}")
print(f"  Unique patients:     {len(set(k[1] for k in prime_keys))}")
print(f"  Unique patient-eyes: {len(set((k[1], k[3]) for k in prime_keys))}")
print(f"  Unique visits:       {len(set(k[2] for k in prime_keys))}")

# Average visits per patient-eye
def visits_per_eye(keys_subset):
    eye_visits = {}
    for k in keys_subset:
        pe = (k[1], k[3])
        eye_visits.setdefault(pe, set()).add(k[2])
    return [len(v) for v in eye_visits.values()]

trex_visits = visits_per_eye(trex_keys)
prime_visits = visits_per_eye(prime_keys)

print(f"\nVisits per eye:")
print(f"  TREX DME -- mean: {sum(trex_visits)/len(trex_visits):.1f}, max: {max(trex_visits)}, eyes: {len(trex_visits)}")
print(f"  Prime    -- mean: {sum(prime_visits)/len(prime_visits):.1f}, max: {max(prime_visits)}, eyes: {len(prime_visits)}")

print(f"\nTotal across both trials:")
print(f"  Total fundus files:  {len(keys)}")
print(f"  Total unique visit-eyes: {len(set(keys))}")
print(f"  Total unique patient-eyes: {len(set((k[0], k[1], k[3]) for k in keys))}")
print(f"  Total unique patients (across trials): {len(set((k[0], k[1]) for k in keys))}")

TREX DME fundus files: 1849
  Unique visit-eyes:   1849
  Unique patients:     56
  Unique patient-eyes: 120
  Unique visits:       27

Prime_FULL fundus files: 2586
  Unique visit-eyes:   1293
  Unique patients:     40
  Unique patient-eyes: 84
  Unique visits:       21

Visits per eye:
  TREX DME -- mean: 15.4, max: 27, eyes: 120
  Prime    -- mean: 15.4, max: 21, eyes: 84

Total across both trials:
  Total fundus files:  4435
  Total unique visit-eyes: 3142
  Total unique patient-eyes: 204
  Total unique patients (across trials): 96


In [ ]:
from google.colab import userdata
TOKEN = userdata.get('GITHUB_TOKEN')  # Reads from Colab Secrets

# 1. Mount Drive (if not already mounted)
from google.colab import drive
drive.mount('/content/drive')

# 2. Re-clone the repo
import os
from google.colab import userdata

REPO_DIR = "/content/dr-dissertation"

if not os.path.exists(REPO_DIR):
    !git clone https://savita10:{TOKEN}@github.com/savita10/dr-dissertation.git {REPO_DIR}
    print("Cloned fresh")
else:
    !git -C {REPO_DIR} pull
    print("Pulled latest")

# 3. Delete stale OLIVES outputs (the early-exit guard would skip otherwise)
!rm -f /content/drive/MyDrive/dissertation/preprocessed/olives/olives_fundus.pt
!rm -f /content/drive/MyDrive/dissertation/preprocessed/olives/metadata.json

# 4. Run OLIVES preprocessing
!cd /content/dr-dissertation && python -m src.data.preprocess --dataset olives

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Cloning into '/content/dr-dissertation'...
remote: Enumerating objects: 120, done.
remote: Counting objects: 100% (120/120), done.
remote: Compressing objects: 100% (75/75), done.
remote: Total 120 (delta 58), reused 100 (delta 38), pack-reused 0 (from 0)
Receiving objects: 100% (120/120), 34.83 KiB | 1.34 MiB/s, done.
Resolving deltas: 100% (58/58), done.
Cloned fresh

OLIVES preprocessing

[1/6 Extracting fundus images to /content/scratch/olives_fundus]
  Opening outer zip: /content/drive/MyDrive/olive.zip
  Opening OLIVES.zip as stream...
  Processing OLIVES/TREX_DME.zip...
    Found 1849 fundus files in TREX_DME
Extracting TREX_DME: 100% 1849/1849 [02:11<00:00, 14.01it/s]
  Processing OLIVES/Prime_FULL.zip...
    Found 2586 fundus files in Prime_FULL
Extracting Prime_FULL: 100% 2586/2586 [01:55<00:00, 22.46it/s]
extraction complete in 2280.6s

[2/6 Loadin

In [ ]:
import torch

data = torch.load(
    "/content/drive/MyDrive/dissertation/preprocessed/olives/olives_fundus.pt",
    weights_only=False,
)

tier_0 = (~data['has_clinical'] & ~data['has_biomarkers']).sum().item()
tier_1 = (data['has_clinical'] & ~data['has_biomarkers']).sum().item()
tier_2 = (data['has_clinical'] & data['has_biomarkers']).sum().item()
biomarker_only = (~data['has_clinical'] & data['has_biomarkers']).sum().item()

print(f"Tier 0 (no labels):           {tier_0}")
print(f"Tier 1 (clinical only):       {tier_1}")
print(f"Tier 2 (clinical + biomarker): {tier_2}")
print(f"Biomarker-only (anomalous):   {biomarker_only}")
print(f"Total:                        {tier_0 + tier_1 + tier_2 + biomarker_only}")

# Re-verify biomarker positive counts (should match what we had before, within tier-2 sample count)
biomarker_names = [
    'Atrophy/thinning', 'EZ Disruption', 'DRIL', 'IR hemorrhages',
    'IR HRF', 'Vitreous (partial)', 'Vitreous (full)', 'Preretinal',
    'Vitreous debris', 'VMT', 'DRT/ME', 'IRF', 'SRF', 'RPE Disruption',
    'PED (serous)', 'SHRM'
]

biomarkers = data['biomarkers'][data['has_biomarkers']]
print(f"\nBiomarker positive counts (out of {len(biomarkers)} tier-2 samples):")
print(f"  {'Biomarker':<25s} {'Pos':>5s} {'%':>6s}")
for i, name in enumerate(biomarker_names):
    col = biomarkers[:, i]
    valid = col[~torch.isnan(col)]
    pos = (valid > 0).sum().item()
    pct = pos / len(valid) * 100 if len(valid) > 0 else 0
    print(f"  {name:<25s} {pos:>5d} {pct:>5.1f}%")

Tier 0 (no labels):           1546
Tier 1 (clinical only):       1404
Tier 2 (clinical + biomarker): 190
Biomarker-only (anomalous):   2
Total:                        3142

Biomarker positive counts (out of 192 tier-2 samples):
  Biomarker                   Pos      %
  Atrophy/thinning             17   8.9%
  EZ Disruption                68  35.4%
  DRIL                          5   2.6%
  IR hemorrhages               67  34.9%
  IR HRF                      192 100.0%
  Vitreous (partial)          110  57.3%
  Vitreous (full)             134  69.8%
  Preretinal                   48  25.0%
  Vitreous debris             164  85.4%
  VMT                           3   1.6%
  DRT/ME                      114  59.4%
  IRF                         174  90.6%
  SRF                          29  15.1%
  RPE Disruption                6   3.1%
  PED (serous)                  2   1.0%
  SHRM                         17   8.9%


# Scratch backup utility

Cell below copies the extracted fundus images from Colab's local scratch to a persistent location on Drive. This avoids re-extracting the 4,435 fundus files (1-2 hours from nested zip) in future sessions.


In [ ]:
#back up scratch to Drive
import os
import shutil
from pathlib import Path

# Source: extracted fundus images on Colab local disk (will be wiped on session disconnect)
src = Path("/content/scratch/olives_fundus")

# Destination: persistent location on Drive
dst_root = Path("/content/drive/MyDrive/dissertation/scratch_backup")
dst = dst_root / "olives_fundus"

# Verify source exists and has the expected count
if not src.exists():
    print(f"ERROR: Source {src} does not exist — nothing to back up")
else:
    fundus_files = list(src.rglob("*fundus*"))
    print(f"Source has {len(fundus_files)} fundus files (expect ~4,435 — pre-dedup count)")

    if len(fundus_files) < 4000:
        print(f"WARNING: file count is lower than expected — backup may be incomplete")

    # Create destination
    dst_root.mkdir(parents=True, exist_ok=True)

    # Remove any prior backup to avoid mixing old/new files
    if dst.exists():
        print(f"Removing old backup at {dst}...")
        shutil.rmtree(dst)

    print(f"Copying {src} -> {dst} (this will take ~10-20 minutes for ~4,435 files)...")
    shutil.copytree(src, dst)

    # Verify
    backed_up = list(dst.rglob("*fundus*"))
    print(f"Backup complete: {len(backed_up)} fundus files at {dst}")

ERROR: Source /content/scratch/olives_fundus does not exist — nothing to back up


In [ ]:
from google.colab import drive
drive.mount('/content/drive', force_remount=True)

# Check again
import os
path = "/content/drive/MyDrive/dissertation/preprocessed/olives/olives_fundus.pt"
print(f"Exists after remount: {os.path.exists(path)}")

Mounted at /content/drive
Exists after remount: True


# Final validation suite

Comprehensive validation of the preprocessed OLIVES dataset — file integrity, tier counts, mask consistency, biomarker positive counts. Used as the gate before declaring preprocessing complete.


In [ ]:
import torch
import os
from pathlib import Path
from datetime import datetime

PT_PATH = Path("/content/drive/MyDrive/dissertation/preprocessed/olives/olives_fundus.pt")
META_PATH = Path("/content/drive/MyDrive/dissertation/preprocessed/olives/metadata.json")

print("=" * 70)
print("OLIVES preprocessed dataset validation")
print("=" * 70)

# 1. File existence and metadata
print("\n[1/5 File existence check]")
if not PT_PATH.exists():
    print(f"  ✗ MISSING: {PT_PATH}")
    print("  → You need to re-run preprocessing")
else:
    size_mb = PT_PATH.stat().st_size / 1024 / 1024
    mtime = datetime.fromtimestamp(PT_PATH.stat().st_mtime)
    print(f"  ✓ Found: {PT_PATH}")
    print(f"    Size: {size_mb:.1f} MB")
    print(f"    Last modified: {mtime.isoformat()}")

if META_PATH.exists():
    import json
    with open(META_PATH) as f:
        meta = json.load(f)
    print(f"\n  metadata.json contents:")
    print(f"    Total images: {meta.get('total_images', 'missing')}")
    print(f"    Preprocessing date: {meta.get('preprocessing_date', 'missing')}")
    if 'olives_label_coverage' in meta:
        cov = meta['olives_label_coverage']
        print(f"    Label coverage: {cov.get('total_fundus')} fundus, "
              f"{cov.get('with_clinical_labels')} clinical, "
              f"{cov.get('with_biomarker_labels')} biomarker")

if not PT_PATH.exists():
    raise SystemExit("Cannot validate further without the .pt file")

# 2. Load and check schema
print("\n[2/5 Schema validation]")
data = torch.load(PT_PATH, map_location="cpu", weights_only=False)

required_keys = {
    'images', 'labels', 'metadata', 'biomarkers', 'bcva', 'cst',
    'has_clinical', 'has_biomarkers', 'patient_ids', 'eye_ids', 'keys'
}
present_keys = set(data.keys())
missing = required_keys - present_keys
extra = present_keys - required_keys

if missing:
    print(f"  ✗ MISSING keys: {missing}")
    print("  → Your .pt file is from an older preprocessing version. Re-run preprocessing.")
else:
    print(f"  ✓ All 11 expected keys present")
if extra:
    print(f"  Note: extra keys (informational only): {extra}")

# 3. Shape and count check
print("\n[3/5 Tensor shape validation]")
n = len(data['images'])
print(f"  Total samples: {n}")

expected_shapes = {
    'images': (n, 3, 224, 224),
    'labels': (n,),
    'biomarkers': (n, 16),
    'bcva': (n,),
    'cst': (n,),
    'has_clinical': (n,),
    'has_biomarkers': (n,),
    'patient_ids': (n,),
    'eye_ids': (n,),
}

all_shapes_ok = True
for key, expected in expected_shapes.items():
    actual = tuple(data[key].shape)
    ok = actual == expected
    marker = "✓" if ok else "✗"
    print(f"  {marker} {key:18s} shape={actual}  (expected {expected})")
    all_shapes_ok &= ok

list_keys_lens = {k: len(data[k]) for k in ['metadata', 'keys']}
for key, length in list_keys_lens.items():
    ok = length == n
    marker = "✓" if ok else "✗"
    print(f"  {marker} {key:18s} len={length}  (expected {n})")

# 4. Which version of preprocessing is this from?
print("\n[4/5 Preprocessing version detection]")
if n == 4435:
    print(f"  ⚠ {n} samples — this is from the PRE-DEDUP run (before .tif/.png fix)")
    print("    Latest preprocessing produces 3,142 samples after .tif/.png dedup")
    print("    You may want to re-run preprocessing to get the cleaner deduplicated dataset")
elif n == 3142:
    print(f"  ✓ {n} samples — this is from the LATEST preprocessing (post-dedup)")
elif n == 272:
    print(f"  ⚠ {n} samples — this is from the ORIGINAL broken matching (pre-three-tier refactor)")
    print("    You MUST re-run preprocessing — this version is missing tier-1/tier-0 samples")
else:
    print(f"  ? {n} samples — unrecognised count, expected 272, 4435, or 3142")

# 5. Tier breakdown sanity check
print("\n[5/5 Tier breakdown]")
tier_0 = (~data['has_clinical'] & ~data['has_biomarkers']).sum().item()
tier_1 = (data['has_clinical'] & ~data['has_biomarkers']).sum().item()
tier_2 = (data['has_clinical'] & data['has_biomarkers']).sum().item()
biomarker_only = (~data['has_clinical'] & data['has_biomarkers']).sum().item()
total = tier_0 + tier_1 + tier_2 + biomarker_only

print(f"  Tier 0 (no labels):           {tier_0}")
print(f"  Tier 1 (clinical only):       {tier_1}")
print(f"  Tier 2 (clinical + biomarker): {tier_2}")
print(f"  Biomarker-only (anomalous):   {biomarker_only}")
print(f"  Total:                        {total}")
print(f"  Reconciliation: {'✓ matches' if total == n else '✗ MISMATCH'}")

# Quick biomarker sanity check (constant biomarker indicates broken dedup)
print("\n  Biomarker check (any constant column = broken dedup):")
biomarkers = data['biomarkers'][data['has_biomarkers']]
biomarker_names = [
    'Atrophy/thinning', 'EZ Disruption', 'DRIL', 'IR hemorrhages',
    'IR HRF', 'Vit (partial)', 'Vit (full)', 'Preretinal',
    'Vit debris', 'VMT', 'DRT/ME', 'IRF', 'SRF', 'RPE', 'PED', 'SHRM'
]
zero_count_cols = []
for i, name in enumerate(biomarker_names):
    col = biomarkers[:, i]
    valid = col[~torch.isnan(col)]
    pos = (valid > 0).sum().item()
    if pos == 0:
        zero_count_cols.append(name)
if len(zero_count_cols) > 4:
    print(f"  ⚠ {len(zero_count_cols)} biomarkers have ZERO positives:")
    print(f"    {zero_count_cols}")
    print(f"    This suggests the OLD broken dedup (pre-max-aggregation fix)")
else:
    print(f"  ✓ Biomarker positives look reasonable (only {len(zero_count_cols)} zero columns)")

print("\n" + "=" * 70)
print("Validation complete")
print("=" * 70)
# 6. Direct dedup verification — check for .tif/.png pairs in metadata
print("\n[6/5 .tif/.png dedup verification]")
from collections import Counter

# Count how many fundus files exist per visit-eye key
key_counts = Counter(data['keys'])
duplicates = {k: c for k, c in key_counts.items() if c > 1}

if not duplicates:
    print(f"  ✓ No duplicate visit-eye keys — .tif/.png dedup IS APPLIED")
else:
    n_dup_keys = len(duplicates)
    n_dup_files = sum(c - 1 for c in duplicates.values())
    print(f"  ✗ Found {n_dup_keys} visit-eye keys with multiple files ({n_dup_files} duplicate files total)")
    print(f"  → .tif/.png dedup NOT APPLIED — this is from the pre-dedup preprocessing run")
    print(f"\n  Sample duplicates (first 3):")
    for key, count in list(duplicates.items())[:3]:
        print(f"    {key}: {count} files")

# Cross-check: per-trial sample counts should match expected post-dedup numbers
print(f"\n  Per-trial breakdown:")
trex_count = sum(1 for k in data['keys'] if k[0] == 'TREX DME')
prime_count = sum(1 for k in data['keys'] if k[0] == 'Prime_FULL')
print(f"    TREX DME: {trex_count}  (expected 1849 — same pre/post dedup)")
print(f"    Prime_FULL: {prime_count}  (expected 1293 post-dedup, 2586 pre-dedup)")

if trex_count == 1849 and prime_count == 1293:
    print(f"  ✓ Per-trial counts match POST-DEDUP expectations")
elif trex_count == 1849 and prime_count == 2586:
    print(f"  ✗ Per-trial counts match PRE-DEDUP — re-run preprocessing for the cleaner dataset")
else:
    print(f"  ? Per-trial counts don't match either expected pattern — investigate")

OLIVES preprocessed dataset validation

[1/5 File existence check]
  ✓ Found: /content/drive/MyDrive/dissertation/preprocessed/olives/olives_fundus.pt
    Size: 1804.8 MB
    Last modified: 2026-04-27T21:35:19

  metadata.json contents:
    Total images: 3142
    Preprocessing date: 2026-04-27T21:35:19.663279+00:00
    Label coverage: 3142 fundus, 1594 clinical, 192 biomarker

[2/5 Schema validation]
  ✓ All 11 expected keys present

[3/5 Tensor shape validation]
  Total samples: 3142
  ✓ images             shape=(3142, 3, 224, 224)  (expected (3142, 3, 224, 224))
  ✓ labels             shape=(3142,)  (expected (3142,))
  ✓ biomarkers         shape=(3142, 16)  (expected (3142, 16))
  ✓ bcva               shape=(3142,)  (expected (3142,))
  ✓ cst                shape=(3142,)  (expected (3142,))
  ✓ has_clinical       shape=(3142,)  (expected (3142,))
  ✓ has_biomarkers     shape=(3142,)  (expected (3142,))
  ✓ patient_ids        shape=(3142,)  (expected (3142,))
  ✓ eye_ids            s

---

# Summary

## What was produced

**EyePACS** preprocessed dataset (saved to `dissertation/preprocessed/eyepacs/` on Drive):
- 35,108 fundus images at 3×224×224 float32, ImageNet-normalised
- 8 shard files (~22 GB total)
- 5-class DR severity labels (No DR through Proliferative DR)

**OLIVES** preprocessed dataset (saved to `dissertation/preprocessed/olives/` on Drive):
- 3,142 unique fundus images at 3×224×224 float32, ImageNet-normalised (~1.8 GB)
- Three-tier label structure with explicit per-tier masks:
  - Tier 0 (no labels): 1,546 samples (49.2%)
  - Tier 1 (BCVA/CST clinical labels): 1,404 samples (44.7%)
  - Tier 2 (full biomarker labels): 190 samples (6.0%)
  - Anomalous (biomarker-only edge case): 2 samples
- 96 patients, 204 patient-eyes — matches the OLIVES paper's reported cohort exactly

## Bugs found and fixed during this notebook

1. **Clinical xlsx column name mismatch** — code expected `Path (Trial/Arm/Folder/Visit/Eye/Image Name)` but clinical file uses `File_Path`. Fixed by introducing separate column name constants.

2. **Biomarker deduplication losing positives** — `drop_duplicates(keep='first')` silently dropped 60-70% of positives. Replaced with `groupby().max()` (any-positive aggregation rule).

3. **Prime_FULL `.tif`/`.png` duplication** — same fundus image stored in both formats, causing 2× inflation. Fixed by preferring `.tif` over `.png` for the same source image.

## What to do next

1. Verify the final cell above (cell 51) shows all green checks for the OLIVES validation suite.
2. Move to `03_phase1_distributional_validation.ipynb` to run the six distributional diagnostics on these preprocessed datasets.
3. The preprocessing decisions documented here should be referenced in the dissertation methods chapter — see `docs/PHASE1_SUMMARY.md` Section 6 for the formal write-up.
